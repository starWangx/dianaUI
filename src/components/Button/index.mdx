---
name: Button —— 按钮
route: /button
---
import { Playground, Props } from 'docz'
import Button from './index.tsx'

# Button 按钮组件

<Props of={Button} />

## Demo

<Playground>

<table>
  <tr>
    <td></td>
    <td align="center" >Normal</td>
    <td align="center" >Disable</td>
  </tr>

  <tr height='80px'>
    <td width='120px' > Primary Button </td>
    <td width='120px' align="center"> 
      <Button
        theme='primary'
        >
        主按钮
      </Button>
    </td>
    <td align="center"> 
      <Button
        theme='primary'
        disabled
        >
        主按钮
      </Button>
    </td>
    
  </tr>


  <tr height='80px'>
    <td>Default Button </td>
    <td align="center"> 
      <Button
        theme='default'
        >
        次按钮
      </Button>
    </td>
    <td align="center"> 
      <Button
        theme='default'
        disabled
        >
        次按钮
      </Button>
    </td>
  </tr>

  <tr height='80px'>
    <td>Icon Button</td>
    <td align="center"> 
      <Button
        theme='icon'
        >
        保存
      </Button>
    </td>
    <td align="center"> 
      <Button
        theme='icon'
        disabled
        >
        保存
      </Button>
    </td>
  </tr>

  <tr height='80px'>
    <td>Menu Button</td>
    <td align="center"> 
      <Button
        theme='menu'
        size='small'
        >
        small
      </Button>
    </td>
    <td align="center"> 
      <Button
        theme='menu'
        disabled
        >
        菜单按钮
      </Button>
    </td>
  </tr>

</table>
    
</Playground>


## API - Props

#### `theme` 四种主题: `Primary`( 主按钮 )、`Default`( 次按钮 )、`icon`( 含icon )、`menu`( 菜单按钮 )
#### `size` 三种型号: `small`( 小 )、`default`( 默认 )、`large`( 大 )
#### `disabled` 是否可点