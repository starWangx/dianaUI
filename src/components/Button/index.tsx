import React, { SFC, ReactNode } from 'react';
import { Button, ButtonProps } from '@hupu/rc-basic';  // 引入公共组件库
import styled from 'styled-components'; // styled-components

// 主题颜色
import * as C from '../../styles/theme';

// 可控属性
type Size = 'small' | 'large' | 'default';
type Theme = 'primary' | 'default' | 'icon' | 'menu';

// 组件主题传参
type CssProps = { 
  disabled: boolean | void;
  size: Size;
  theme: Theme;
 };

export interface ButtonUIProps extends ButtonProps { 
  size?: Size;
  theme?: Theme;
  disabled?: boolean;
 };

function themeMapToCss (theme: Theme, property: string, tag?: string): string { 
  let cssStr = '';

  switch (property) { 
  case 'border':
    if ( theme === 'primary' ) cssStr = `1px solid ${ C.colorBlue1A68FF }`;
    else if ( theme === 'default' ) cssStr = `1px solid ${ C.colorBlue1A68FF }`;
    else if ( theme === 'icon' ) cssStr = `1px solid ${ C.colorGrayd9d9d9 }`;
    else if ( theme === 'menu' ) cssStr = `1px solid ${ C.colorGrayd9d9d9 }`;
    break;
  case 'color':
    if ( theme === 'primary' ) cssStr =`${ C.colorwhitefff }`;
    else if ( theme === 'default' ) cssStr =`${ C.colorBlue1A68FF }`;
    else if ( theme === 'icon' ) cssStr =`${ C.colorGray666 }`;
    else if ( theme === 'menu' ) cssStr =`${ C.colorGray666 }`;
    break;
  case 'bg':
    if ( theme === 'primary' ) cssStr = `${ C.colorBlue1A68FF }`;
    else if ( theme === 'default' ) cssStr = `${ C.colorwhitefff }`;
    else if ( theme === 'icon' ) cssStr = `${ C.colorwhitefff }`;
    else if ( theme === 'menu' ) cssStr = `${ C.colorwhitefff }`;
    break;
  }

  return cssStr;
}


function sizeMapToCss (size: Size, property: string): string { 
  let cssStr = '';

  switch ( property) { 
  case 'width':
    if ( size === 'small' ) cssStr = '60px';
    else if ( size === 'default' ) cssStr = '80px';
    else if ( size === 'large' ) cssStr = '85px';
    break;
  case 'height':
    if ( size === 'small' ) cssStr = '30px';
    else if ( size === 'default' ) cssStr = '34px';
    else if ( size === 'large' ) cssStr = '42px';
    break;
  case 'lineheight':
    if ( size === 'small' ) cssStr = '28px';
    else if ( size === 'default' ) cssStr = '32px';
    else if ( size === 'large' ) cssStr = '40px';
    break;
  case 'fz':
    if ( size === 'small' ) cssStr = '12px';
    else if ( size === 'default' ) cssStr = '14px';
    else if ( size === 'large' ) cssStr = '16px';
    break;
  }

  return cssStr;
}

const Wrapper = styled.div<CssProps>`
  display: inline-block;
  border-radius: 4px;
  box-sizing: border-box;
  text-align:center;
  font-family: PingFangSC-Regular;

  cursor: ${ props => props.disabled ? 'not-allowed' :'pointer' };
  background: ${ props => props.disabled ? `${ C.colorGrayf7f7f7 }` : themeMapToCss( props.theme, 'bg') };
  border: ${ props => props.disabled ? `${ C.colorGrayd9d9d9 }` : themeMapToCss( props.theme, 'border' ) };    
  color: ${ props => props.disabled ? `${ C.colorGrayccc }`: themeMapToCss( props.theme, 'color' ) };

  width: ${ props => sizeMapToCss( props.size, 'width' ) };
  height: ${ props => sizeMapToCss( props.size, 'height' ) };    
  line-height: ${ props => sizeMapToCss( props.size, 'lineheight'  ) };   
  font-size: ${ props => sizeMapToCss( props.theme, 'fz' ) }; 
`;

const StyledButton = styled(Button)<ButtonUIProps>``;

const ButtonUI: SFC<ButtonUIProps> = props => { 
  const {  children, size = 'default', theme , disabled, ...rest  } = props;

  return (
    <Wrapper
      disabled={ disabled }
      theme={ theme }
      size={ size }
    >
      <StyledButton
        {  ...rest  }
      >
        {  children  }
      </StyledButton>
    </Wrapper>
  );
};

export default ButtonUI;