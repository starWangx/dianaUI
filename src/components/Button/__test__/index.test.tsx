import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Button from '../index';

configure({ adapter: new Adapter() });

describe('Button', () => {
  it('renders correctly - shallow', () => {
    const wrapper = shallow(<Button />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - render', () => {
    const wrapper = render(<Button />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - mount', () => {
    const wrapper = mount(<Button />);

    expect(wrapper).toMatchSnapshot();
  });
});
