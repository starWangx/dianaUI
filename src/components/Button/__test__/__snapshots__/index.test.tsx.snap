// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly - mount 1`] = `ReactWrapper {}`;

exports[`Button renders correctly - render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bdVaJa iUObxG",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "default sc-bwzfXH uTxCW",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {},
          "children": Array [],
          "name": "style",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "style",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {},
        "children": Array [],
        "name": "style",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "default sc-bwzfXH uTxCW",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "style",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button renders correctly - shallow 1`] = `ShallowWrapper {}`;
